pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java

```diff
@@ -257,47 +452,198 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
   }
 
   /**
-   * Check if IdealState = ExternalView. If its not equal, return the number of differing segments.
+   * Check if external view has converged to ideal state
+   * @param tableName name of table that we are rebalancing
+   * @return true if external view is same as ideal state, false otherwise
    */
-  public int isStable(String tableName) {
+  private boolean isStable(String tableName) {
     IdealState idealState = _helixAdmin.getResourceIdealState(_helixClusterName, tableName);
     ExternalView externalView = _helixAdmin.getResourceExternalView(_helixClusterName, tableName);
     Map<String, Map<String, String>> mapFieldsIS = idealState.getRecord().getMapFields();
     Map<String, Map<String, String>> mapFieldsEV = externalView.getRecord().getMapFields();
-    int numDiff = 0;
+
+    LOGGER.info("Checking if ideal state and external view are same for table {}", tableName);
+    int segmentsNotConverged = 0;
+
+    boolean stable = true;
     for (String segment : mapFieldsIS.keySet()) {
       Map<String, String> mapIS = mapFieldsIS.get(segment);
       Map<String, String> mapEV = mapFieldsEV.get(segment);
+      boolean converged = true;
+
+      if (mapEV == null) {
+        LOGGER.info("Host-state mapping of segment {} not yet available in external view", segment);
+        // we have found that external view hasn't yet converged to ideal state.
+        // still go on to check for other segments just so that we can dump debug
+        // messages or we can detect error. that's why we don't return
+        // false immediately
+        stable = false;
+        ++segmentsNotConverged;
+        continue;
+      }
+
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Hosts and states for segment {} in ideal state", segment);
+        prettyPrintMap(mapIS, Level.DEBUG);
+        LOGGER.debug("Hosts and states for segment {} in external view", segment);
+        prettyPrintMap(mapEV, Level.DEBUG);
+      }
 
       for (String server : mapIS.keySet()) {
-        String state = mapIS.get(server);
-        if (mapEV == null || mapEV.get(server) == null || !mapEV.get(server).equals(state)) {
-          LOGGER.debug("Mismatch: segment" + segment + " server:" + server + " state:" + state);
-          numDiff = numDiff + 1;
+        if (!mapEV.containsKey(server)) {
+          LOGGER.info("Host-state mapping of segment {} doesn't yet have server {} in external view",
```

Comment:
I think that we should log `debug` level only in this loop. You're logging `num segments * replication` lines of log. This can easily be equivalent to thousands of lines of logs.

Reply:
Only for the segments and servers where external view hasn't converge. Right?

Reply:
oh i missed the if check :)